fv3core/stencils/d_sw.py

```diff
@@ -449,48 +516,39 @@ def get_column_namelist(config: DGridShallowWaterLagrangianDynamicsConfig, npz):
 
 
 @gtscript.function
-def ubke(uc, vc, cosa, rsina, ut, ub, dt4, dt5):
-    from __externals__ import i_end, i_start, j_end, j_start
-
-    ub = dt5 * (uc[0, -1, 0] + uc - (vc[-1, 0, 0] + vc) * cosa) * rsina
-    # if __INLINED(spec.namelist.grid_type < 3):
-    with horizontal(region[:, j_start], region[:, j_end + 1]):
-        ub = dt4 * (-ut[0, -2, 0] + 3.0 * (ut[0, -1, 0] + ut) - ut[0, 1, 0])
-    with horizontal(region[i_start, :], region[i_end + 1, :]):
-        ub = dt5 * (ut[0, -1, 0] + ut)
-
-    return ub
-
-
-def mult_ubke(
-    vb: FloatField,
-    ke: FloatField,
-    uc: FloatField,
-    vc: FloatField,
-    cosa: FloatFieldIJ,
-    rsina: FloatFieldIJ,
-    ut: FloatField,
-    ub: FloatField,
-    dt4: float,
-    dt5: float,
+def interpolate_uc_vc_to_cell_corners(
+    uc_cov, vc_cov, cosa, rsina, uc_contra, vc_contra
 ):
-    with computation(PARALLEL), interval(...):
-        ke = vb * ub
-        ub = ubke(uc, vc, cosa, rsina, ut, ub, dt4, dt5)
-
-
-@gtscript.function
-def vbke(vc, uc, cosa, rsina, vt, vb, dt4, dt5):
+    """
+    Convert covariant C-grid winds to contravariant B-grid (cell-corner) winds.
+    """
     from __externals__ import i_end, i_start, j_end, j_start
 
-    vb = dt5 * (vc[-1, 0, 0] + vc - (uc[0, -1, 0] + uc) * cosa) * rsina
+    # TODO: ask Lucas why we interpolate then convert to contravariant in tile center,
+    # but convert to contravariant and then interpolate on tile edges.
+    ub_cov = 0.5 * (uc_cov[0, -1, 0] + uc_cov)
+    vb_cov = 0.5 * (vc_cov[-1, 0, 0] + vc_cov)
+    ub_contra = contravariant(ub_cov, vb_cov, cosa, rsina)
+    vb_contra = contravariant(vb_cov, ub_cov, cosa, rsina)
     # ASSUME : if __INLINED(namelist.grid_type < 3):
+    with horizontal(region[:, j_start], region[:, j_end + 1]):
+        ub_contra = 0.25 * (
+            -uc_contra[0, -2, 0]
+            + 3.0 * (uc_contra[0, -1, 0] + uc_contra)
+            - uc_contra[0, 1, 0]
+        )
+    with horizontal(region[i_start, :], region[i_end + 1, :]):
+        ub_contra = 0.5 * (uc_contra[0, -1, 0] + uc_contra)
     with horizontal(region[i_start, :], region[i_end + 1, :]):
-        vb = dt4 * (-vt[-2, 0, 0] + 3.0 * (vt[-1, 0, 0] + vt) - vt[1, 0, 0])
+        vb_contra = 0.25 * (
+            -vc_contra[-2, 0, 0]
+            + 3.0 * (vc_contra[-1, 0, 0] + vc_contra)
+            - vc_contra[1, 0, 0]
+        )
```

Comment:
By removing 'dt5' and dt4 from the way this is calculated, this is more difficult to line up what is happening here with the fortran code. I realize you did this so the words you used with xtp_u and ytp_v could be simple things like 'wind' instead of dt * wind, and making sense of xtp_u and ytp_v is the crux of this PR. But this is a lot of mental gymnastics to put on someone trying to port other options or compare the fortran code to the python code (e.g. if debugging a strange behavior present in python but not fortran). Can xtp_u and ytp_v not still be explained in a similar way, with the dt factor included? I  would be more comfortable with just removing dt4 and dt5 variables in favor of explicitly using '0.5 * dt', '0.25 * dt' in their places here, but that's not the point, I know.  Variables including or not including a 'dt' factor does not drastically change our physical understanding of what's happening.  I am surprised this didn't result in validation errors. It's a small thing, but I think we should be careful. At the very minimum the docstring should include something about how this differs from the calculation of ub and vb in the Fortran code.

Reply:
> At the very minimum the docstring should include something about how this differs from the calculation of ub and vb in the Fortran code.

I think this is a very reasonable minimum, and I'll at least do this.

At the namelist options meeting this week, we should discuss our relative prioritization of a) making it easy to understand our model code, and b) making it easy to map between our model code and the Fortran code. We can satisfy both to a certain extent, but these two priorities are inherently at odds with each other in cases where the Fortran code is hard to follow. We could also get feedback from Lucas on this point.

Reply:
I somewhat disagree, I don't think having dt folded in or not is more or less understandable, when it is documented what the variables are. It's super common in atmosphere science to move terms around in equations to help drop pieces off or solve them with different methods. It didn't seem impossible xtp_u/ytp_v could be stepping something forward in time, we just didn't have information on the intention of those subroutines, nor did we spend sufficient time exploring the context in which these subroutines were called when we did the first round of docstrings. But your in-depth investigation has helped us understand, so we do have the capability to document it in a way that makes sense.